src/models/backup.ts

```diff
@@ -460,3 +459,204 @@ export class Drive implements BackupProvider {
+                  localStorage.oneDriveRevoked = true;
+                }
+                console.error(res.error_description);
+                resolve(false);
+              } else {
+                localStorage.oneDriveToken = res.access_token;
+                resolve(true);
+              }
+            } catch (error) {
+              console.error(error);
+              reject(error);
+            }
+          }
+          return;
+        };
+        xhr.send(
+          `client_id=${getCredentials().onedrive.client_id}&refresh_token=${
+            localStorage.oneDriveRefreshToken
+          }&client_secret=${encodeURIComponent(
+            getCredentials().onedrive.client_secret
+          )}&grant_type=refresh_token&scope=https%3A%2F%2Fgraph.microsoft.com%2FFiles.ReadWrite.AppFolder%20https%3A%2F%2Fgraph.microsoft.com%2FUser.Read%20offline_access`
+        );
+      }
+    );
+  }
+
+  async upload(encryption: Encryption) {
+    if (localStorage.oneDriveEncrypted === undefined) {
+      localStorage.oneDriveEncrypted = "true";
+    }
+    const exportData = await EntryStorage.backupGetExport(
+      encryption,
+      localStorage.oneDriveEncrypted === "true"
+    );
+    const backup = JSON.stringify(exportData, null, 2);
+
+    const token = await this.getToken();
+    if (!token) {
+      return new Promise((resolve: (value: boolean) => void) => {
+        resolve(false);
+      });
+    }
+
+    return new Promise(
+      (resolve: (value: boolean) => void, reject: (reason: Error) => void) => {
+        if (!token) {
+          resolve(false);
+        }
+        try {
+          const xhr = new XMLHttpRequest();
+          const now = new Date()
+            .toISOString()
+            .slice(0, 10)
+            .replace(/-/g, "");
+          xhr.open(
+            "PUT",
+            `https://graph.microsoft.com/v1.0/me/drive/special/approot:/${now}.json:/content`
+          );
+          xhr.setRequestHeader("Authorization", "Bearer " + token);
+          xhr.setRequestHeader("Content-type", "application/octet-stream");
+          xhr.onreadystatechange = () => {
+            if (xhr.readyState === 4) {
+              if (xhr.status === 401) {
+                localStorage.removeItem("oneDriveToken");
+                resolve(false);
+              }
+              try {
+                const res = JSON.parse(xhr.responseText);
+                if (!res.error) {
+                  resolve(true);
+                } else {
+                  console.error(res.error.message);
+                  resolve(false);
+                }
+              } catch (error) {
+                reject(error);
+              }
+            }
+            return;
+          };
+          xhr.send(backup);
+        } catch (error) {
+          return reject(error);
+        }
+      }
+    );
+  }
+
+  async getUser() {
+    const token = await this.getToken();
+    if (!token) {
+      return new Promise((resolve: (value: string) => void) => {
```

Comment:
Should we resolve the promise here? It seems reject makes more sense here. Also, here is safe to return the string directly.